Power Law Fit Function for Muon Modelling

sf1919 opened the pull request:
**Description of work.**
Added a new Power Law Fit function for Muon modelling and associated documentation.

Also added links to Gaussian, Lorenztian and Polynomial Fitting functions under MuonModelling of the fitting function tree as part of the WIMDA project.


**To test:**

TBC

Fixes #32665 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Initial framework for PowerLaw Function

Created header and soruce files and added to CMakeLists

sf1919 committed:
Corrected functionDeriv1D method

sf1919 committed:
Updated MuonModelling category in Fitting Tree

Added Gaussian, Lorentzian and Polynomial fitting functions under MuonModelling category

sf1919 committed:
Added documentation

Added release notes and documentation for the new Power Law function.

AnthonyLim23 commented on Framework/CurveFitting/inc/MantidCurveFitting/Functions/PowerLaw.h:
```diff
@@ -0,0 +1,43 @@
+  std::string name() const override { return "PowerLaw"; }
+  void function1D(double *out, const double *xValues, const size_t nData) const override;
+  void functionDeriv1D(API::Jacobian *out, const double *xValues, const size_t nData) override;
+  const std::string category() const override { return "Muon\\MuonModelling"; }
```
might be worth asking in tech QA if anyone else would like this? I can see it being in "general"

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/PowerLaw.cpp:
```diff
@@ -0,0 +1,53 @@
+
+DECLARE_FUNCTION(PowerLaw)
+
+void PowerLaw::init() {
```
can we have some better defaults?

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/PowerLaw.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+void PowerLaw::function1D(double *out, const double *xValues, const size_t nData) const {
+  const double a = getParameter("a");
```
can we have some better parameter names (e.g. magnitude, exponent, constant)?

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/PowerLaw.cpp:
```diff
@@ -0,0 +1,53 @@
+
+  for (size_t i = 0; i < nData; i++) {
+    double diffa = pow(xValues[i], b);
+    double diffb = a * b * pow(xValues[i], b - 1) * log(xValues[i]);
```
The exponent is `b` and not `b-1`. We could also call the `function1D` and the multiply it be the log(x)

AnthonyLim23 commented on docs/source/fitting/fitfunctions/PowerLaw.rst:
```diff
@@ -0,0 +1,23 @@
+Description
+-----------
+
+A power law background function is defined as:
```
lets say something like: "A power law may be used to describe <insert experiment data>"

AnthonyLim23 commented on docs/source/release/v6.3.0/muon.rst:
```diff
@@ -9,4 +9,15 @@ MuSR Changes
+
+New Features
+############
+- Added a :ref:`Power Law <func-PowerLaw>` function to MuonModelling Fit Functions.
```
also worth putting in the framework (see my note above about adding it to general)

sf1919 committed:
Updated parameters, functionDeriv1D and documentation

Updated parameter names and deafults, corrected functionDeriv1D, added PowerLaw to General fitting list, added framework release note, updated PowerLaw documentation.

sf1919 committed:
Added cmath header

cmath header needed for pow and log calculations.

sf1919 committed:
Added and amended tests

Added a test for PowerLaw function. Updated tests for Gaussian and Lorenzian functions as in more than one category. Updated IndirectFitPlotModel to account for Lorenzian being in more than one category.

sf1919 committed:
Capitalise PowerLaw parameters

sf1919 committed:
Updated test and docs after review

Added an example to the documentation and explained parameters. Updated test with improved parameter names.

AnthonyLim23 commented:
there is an error in the derivative if you use the following script, changing between numerical and analytic derivatives gives very different answers. 

```

x =np.linspace(5, 10,200)

def func(x, a,b,c):
    return a +b*(np.power(x,c))

def calc_error(ran):
    return 1. + 0.*(ran-0.5)

def error(f):
    return [y*calc_error(np.random.random()) for y in f]
    return 

constant = 2.5
mag = 1.2
expon = 3.56

param=[mag, expon, constant]

y = error(func(x,constant,mag,expon))
ws = CreateWorkspace(x,y)

f = PowerLaw()#Magnitude=1, Constant=0, Exponent=1)
c = CompositeFunction(Gaussian(PeakCentre=1))
myFunc = CompositeFunction(f, NumDeriv=True)


fit_output = Fit(InputWorkspace=ws, Output='fit', Function=myFunc)
paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters

print("The fit was: {}".format(fit_output.OutputStatus))
for j in range(len(paramTable)-1):
    print(paramTable.column(0)[j], "is", paramTable.column(1)[j]," error is", 100*(paramTable.column(1)[j]-param[j])/param[j])


```

sf1919 committed:
Revision to docs and functionDeriv1D

Following on from review comments the documentation for PowerLaw was extended to include an example. An error in the functionDeriv1D function was also fixed.

AnthonyLim23 reviewed (dismissed):
Tried a few different data sets and the fits converge nicely. For one case the analytic derivatives led to a better set of parameters.

DavidFair reviewed (approved):
LGTM lets :shipit:
